go/worker/keymanager/worker.go

```diff
@@ -306,6 +297,88 @@ func (w *Worker) localCallEnclave(method string, args interface{}, rsp interface
 	return nil
 }
 
+func (w *Worker) authorize(method string, kind enclaverpc.Kind, peerID core.PeerID) error {
+	select {
+	case <-w.initCh:
+	default:
+		return fmt.Errorf("not initialized")
+	}
+
+	switch kind {
+	case enclaverpc.KindInsecureQuery:
+		switch method {
+		case api.RPCMethodGetPublicKey:
+		case api.RPCMethodGetPublicEphemeralKey:
+		default:
+			return fmt.Errorf("unsupported method: %s", method)
+		}
+		return nil
+	case enclaverpc.KindNoiseSession:
+		switch method {
+		case api.RPCMethodGetOrCreateKeys:
+		case api.RPCMethodGetOrCreateEphemeralKeys:
+		case api.RPCMethodReplicateMasterSecret:
+		case api.RPCMethodReplicateEphemeralSecret:
+		default:
+			return fmt.Errorf("unsupported method: %s", method)
+		}
+	default:
+		return fmt.Errorf("unsupported kind: %s", kind)
+	}
+
+	if _, ok := w.privatePeers[peerID]; ok {
+		return nil
+	}
+
+	w.RLock()
+	defer w.RUnlock()
+
+	if w.globalStatus == nil || !w.globalStatus.IsInitialized || w.activeVersion == nil {
+		return fmt.Errorf("not initialized")
+	}
+
+	switch method {
+	case api.RPCMethodGetOrCreateKeys, api.RPCMethodGetOrCreateEphemeralKeys:
+		capabilityTEE, err := w.GetHostedRuntimeCapabilityTEE()
+		if err != nil {
+			return err
+		}
+
+		switch {
+		case capabilityTEE == nil:
+			// Insecure key manager enclaves can be queried by all runtimes (used for testing).
+			if w.globalStatus.IsSecure {
+				return fmt.Errorf("untrusted hardware")
+			}
+			return nil
+		case capabilityTEE.Hardware == node.TEEHardwareIntelSGX:
+			// Secure key manager enclaves can be queried by runtimes specified in the policy.
+			if w.globalStatus.Policy == nil {
+				return fmt.Errorf("policy not set")
+			}
+			rts := w.accessList.Runtimes(peerID)
+			for _, enc := range w.globalStatus.Policy.Policy.Enclaves { // TODO: Use the right enclave identity.
+				for rtID := range enc.MayQuery {
+					if rts.Contains(rtID) {
+						return nil
+					}
+				}
+			}
+			return fmt.Errorf("query not allowed")
+		default:
+			return fmt.Errorf("unsupported hardware: %s", capabilityTEE.Hardware)
+		}
+	case api.RPCMethodReplicateMasterSecret, api.RPCMethodReplicateEphemeralSecret:
+		// Replication is restricted to peers within the same key manager runtime.
+		if !w.accessList.Runtimes(peerID).Contains(w.runtimeID) {
+			return fmt.Errorf("not a key manager")
+		}
+		return nil
+	}
+
+	panic("unreachable code")
```

Comment:
Why panic here instead of returning an error? This seems dangerous in case something is modified above.

Reply:
Agree, it safer. I usually panic in this case as execution of this part of the code is not what the programmer wanted, so something is seriously wrong.